### PR TITLE
Hardcode lifetime categories

### PR DESCRIPTION
**Pull request by mgehre**
- Make DerefType optional in attributes
- Add __is_gsl_owner/__is_gsl_pointer builtins for tests
- Hard code gsl::Owner/gsl::Pointer for std types

TODO: document the builtins and that the DerefType is optional

### CODE REVIEW

**Comment by mgehre** [comment]
TODO: Add attributes to classes that are not directly in std::, like std::vector<>::iterator et al.

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 1]
```diff
@@ -6036,6 +6036,57 @@ static bool canPassInRegisters(Sema &S, CXXRecordDecl *D,
   return HasNonDeletedCopyOrMove;
 }
 
+/// Add [[gsl::Owner]] and [[gsl::Pointer]] attributes for std:: types.
+/// \pre Record->getIdentifiers() is non-null
+void Sema::addOwnerPointerAttribute(CXXRecordDecl *Record) {
+
+  static llvm::StringSet<> StdOwners{
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaDeclCXX.cpp]: For some reason LLVM folks do not really like static initializers. I windows if we could make this a const member of Sema and init it in the ctor. We could also wait with this change until the reviewers actually want it.

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 2]
```diff
@@ -6036,6 +6036,57 @@ static bool canPassInRegisters(Sema &S, CXXRecordDecl *D,
...
+      "priority_queue",
+      "queue",
+      "regex",
+      "set",
+      "stack",
+      "unordered_set",
+      "unordered_map",
+      "unordered_multiset",
+      "unordered_multimap",
+      "variant",
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaDeclCXX.cpp]: What would be the deref type for variant? I think we should not add it now (since we do not have the gsl::lifetime annotation yet).

**Review by Xazax-hun — COMMENTED** [review]
Thanks, this looks great!

You was quite fast!

Do you think we could start filling the pointee type too? Maybe it is not much work and in the end we might not need to make the type argument optional (yet). I feel a bit bad about making it optional until we support the [[gsl::lifetime]] attribute.

**Comment by Xazax-hun** [comment]
> 
> 
> TODO: Add attributes to classes that are not directly in std::, like std::vector<>::iterator et al.

 The `isInStdNamespace` call should return true for those as well right? Maybe we could look up the iterator typedef in STL containers. But if we need to look member types up, adding the pointee type is not that much extra work.

File: clang/test/SemaCXX/attr-gsl-owner-pointer.cpp [hunk 3]
```diff
@@ -59,9 +70,21 @@ class [[gsl::Owner(int)]] AddTheSameLater{};
 class [[gsl::Owner(int)]] AddTheSameLater;
 
 class [[gsl::Owner()]] OwnerWithEmptyParameterList{};
+static_assert(__is_gsl_owner(OwnerWithEmptyParameterList), "");
 
 class [[gsl::Pointer()]] PointerWithEmptyParameterList{};
+static_assert(__is_gsl_pointer(PointerWithEmptyParameterList), "");
```
Comment (Xazax-hun) [inline clang/test/SemaCXX/attr-gsl-owner-pointer.cpp]: Now it is possible to write metaprograms that make decisions based on a type being a Pointer or Owner :) Of course, people not supposed to use __ names diretly, so this is not really exposed to the user. I wonder what will the reviewers think about this.

Comment (mgehre) [reply clang/test/SemaCXX/attr-gsl-owner-pointer.cpp]: I copied the approach from https://reviews.llvm.org/D50119, which makes testing very clean. Seems to be not controversial there.

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 2 above)
Comment (mgehre) [reply clang/lib/Sema/SemaDeclCXX.cpp]: As we discussed: Without doing automatic function model inference, we don't need the deref type.
Later, we add [[gsl::lifetime]] to the `std::get()` function to make that work.
The lifetime paper explicitly says 
```
Additionally, for convenient adoption without modifying existing standard library headers, the following well-known standard types are treated as-if annotated as Owners: stack, queue, priority_queue, optional, variant, any, and regex.
```

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 1 above)
Comment (mgehre) [reply clang/lib/Sema/SemaDeclCXX.cpp]: Note that this is not a static initializer in the sense that it runs before main - it runs on first entry to the function. I would wait for the review.
I actually would love to make this a constexpr container, but there seems to be no constexpr map in std/llvm yet.

**Comment by mgehre** [comment]
I would guess  that the standard allows std::vector<>::iterator to be implemented in a different namespace (e.g. std::__internal) or as a member class of std::vector.

Do you know where we can put tests of the hard-coded annotations against real STL implementations? The tests in clang/ all need to be independent of the STL. Would https://github.com/llvm-mirror/test-suite be the right place?

**Comment by Xazax-hun** [comment]
> 
> 
> I would guess that the standard allows std::vector<>::iterator to be implemented in a different namespace (e.g. std::__internal) or as a member class of std::vector.

I think the easiest way to get it done is to also do some processing on typedefs/usings and if it is inside a container like vector, and the name of the declaration is iterator or const_iterator than we can mark the desugared type as Pointer. We also need to handle the case if iterator is actually an inline class. 

> 
> Do you know where we can put tests of the hard-coded annotations against real STL implementations? The tests in clang/ all need to be independent of the STL. Would https://github.com/llvm-mirror/test-suite be the right place?

I think it is quite OK to just mock multiple STL interfaces in some headers.

**Comment by mgehre** [comment]
> I think it is quite OK to just mock multiple STL interfaces in some headers.

Yeah, it's okay as a unit test, but it will not prove that our implementation actually works with the MSVC STL, libstdc++, libcxx and other STLs that have a build bot. I remember that we had quite some interesting edge case with the different STLs in our previous work.

**Review by Xazax-hun — COMMENTED** [review]
Wow, nice work!

File: clang/include/clang/Basic/Attr.td [hunk 4]
```diff
@@ -2769,14 +2769,16 @@ def TypeTagForDatatype : InheritableAttr {
 def Owner : InheritableAttr {
   let Spellings = [CXX11<"gsl", "Owner">];
   let Subjects = SubjectList<[CXXRecord]>;
-  let Args = [TypeArgument<"DerefType">];
+  let Args = [TypeArgument<"DerefType", 1>];
```
Comment (Xazax-hun) [inline clang/include/clang/Basic/Attr.td]: Maybe
```
let Args = [TypeArgument<"DerefType", /*opt*/1>];
```
?

File: clang/lib/Sema/SemaAttr.cpp [hunk 5]
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
...
+      "regex",
+      "set",
+      "stack",
+      "unordered_set",
+      "unordered_map",
+      "unordered_multiset",
+      "unordered_multimap",
+  };
+
+  static llvm::StringSet<> Iterators{"iterator", "const_iterator"};
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaAttr.cpp]: Should we add `reverse_iterator` and `const_reverse_iterator` as well?

File: clang/lib/Sema/SemaAttr.cpp [hunk 6]
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
...
+}
+
+void Sema::addDefaultOwnerPointerAttribute(TypedefNameDecl *TD) {
+  addDefaultOwnerPointerAttribute(
+      TD, TD->getUnderlyingType().getCanonicalType()->getAsCXXRecordDecl());
+}
+
+/// Add [[gsl::Owner]] and [[gsl::Pointer]] attributes for std:: types.
+void Sema::addDefaultOwnerPointerAttribute(CXXRecordDecl *Record) {
+  static llvm::StringSet<> StdOwners{
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaAttr.cpp:145]: I wonder if we should reuse the containers dict from above.

File: clang/include/clang/Basic/Attr.td [hunk 7]
```diff
@@ -2769,14 +2769,16 @@ def TypeTagForDatatype : InheritableAttr {
...
+  let MeaningfulToClassTemplateDefinition = 1;
   let Documentation = [LifetimeOwnerDocs];
 }
 
 def Pointer : InheritableAttr {
   let Spellings = [CXX11<"gsl", "Pointer">];
   let Subjects = SubjectList<[CXXRecord]>;
-  let Args = [TypeArgument<"DerefType">];
+  let Args = [TypeArgument<"DerefType", 1>];
+    let MeaningfulToClassTemplateDefinition = 1;
```
Comment (Xazax-hun) [inline clang/include/clang/Basic/Attr.td]: Indentation is off here.

File: clang/lib/Sema/SemaAttr.cpp [hunk 8]
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
...
+    return;
+
+  if (Record->isInStdNamespace()) {
+    CXXRecordDecl *Canonical = Record->getCanonicalDecl();
+    if (Canonical->hasAttr<OwnerAttr>() || Canonical->hasAttr<PointerAttr>())
+      return;
+
+    if (StdOwners.count(Record->getName())) {
+      addOwnerPointerAttributeIfNotExisting<OwnerAttr>(Context, Record);
+    } else if (StdPointers.count(Record->getName())) {
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaAttr.cpp]: Redundant braces.

File: clang/lib/Sema/SemaAttr.cpp [hunk 9]
```diff
@@ -85,6 +85,116 @@ void Sema::AddMsStructLayoutForRecord(RecordDecl *RD) {
...
+
+  Canonical->addAttr(::new (Context) Attribute(SourceRange{}, Context,
+                                               /*DerefType*/ nullptr,
+                                               /*Spelling=*/0));
+}
+
+/// Add gsl::Pointer attribute to std::container::iterator
+/// \param ND The declaration that introduces the name std::container::iterator.
+/// \param UnderlyingRecord The record named by ND.
+void Sema::addDefaultOwnerPointerAttribute(NamedDecl *ND,
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaAttr.cpp]: In case this only ever adds Pointer attribute maybe we should get rid of the Owner from the name?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 10]
```diff
@@ -4550,7 +4553,10 @@ static void handleLifetimeCategoryAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
     if (checkAttrMutualExclusion<PointerAttr>(S, D, AL))
       return;
     if (const auto *OAttr = D->getAttr<OwnerAttr>()) {
-      if (OAttr->getDerefType().getTypePtr() != ParmType.getTypePtr()) {
+      const Type *ExistingDerefType = OAttr->getDerefTypeLoc()
```
Comment (Xazax-hun) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Do we actually need to have the ternary operator? Could we just rely on comparing to QualTypes? Or two invalid QualTypes compared inequal? Then we could use getTypePtrOrNull on both of them.

**Review by Xazax-hun — COMMENTED** [review]
Great work, some nits inline :)

File: clang/lib/Sema/SemaAttr.cpp (hunk 6 above)
Comment (mgehre) [reply clang/lib/Sema/SemaAttr.cpp:145]: They are not exactly the same. `Containers` does only contain classes that have iterators, i.e it does not contain `any`, `variant` and `basic_regex`.

Comment (Xazax-hun) [reply clang/lib/Sema/SemaAttr.cpp:145]: Yeah, but if there is a subset relationship we may be able utilize that. It would mean two lookups instead of one though.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 10 above)
Comment (mgehre) [reply clang/lib/Sema/SemaDeclAttr.cpp]: `getDerefType` dereferences a null pointer when `getDerefTypeLoc` is null.

File: clang/test/SemaCXX/attr-gsl-owner-pointer.cpp [hunk 11]
```diff
@@ -131,7 +131,7 @@ template <
     class CharT,
     class Traits>
 class basic_regex;
-static_assert(__is_gsl_pointer(basic_regex<char, void>), "");
+static_assert(__is_gsl_owner(basic_regex<char, void>), "");
```
Comment (Xazax-hun) [inline clang/test/SemaCXX/attr-gsl-owner-pointer.cpp]: Don't we have a static_assert overload that does not take a second argument?

Comment (mgehre) [reply clang/test/SemaCXX/attr-gsl-owner-pointer.cpp]: Since C++17, yes. Is it safe to change the test to run in C++17 mode? Most tests seem to just use the default language mode. I copied that part also from https://reviews.llvm.org/D50119

Comment (Xazax-hun) [reply clang/test/SemaCXX/attr-gsl-owner-pointer.cpp]: I see, makes sense.

**Comment by Xazax-hun** [comment]
As far as I understand this is not relevant anymore :)
